[Ready] Ports dynamic gamemode from /vg/

TheChosenEvilOne opened the pull request:
## About The Pull Request
Port the dynamic gamemode from /vg/.
(Really bad explanation of the mode incoming.)
The dynamic game mode generates a threat number which is used to "buy" rulesets (rulesets are basically your antagonists). This means you can have rounds with for example traitors and cult (you can have up to three roundstart rulesets depending on the pop and threat level), and then there are latejoin and midround rulesets which basically do what they say (latejoin ruleset assigns late joining player as an antagonists and midround assigns ghosts or a currently alive player as an antagonist)
## Why It's Good For The Game
This increases the chances of people getting their important antagonist role and makes rounds more interesting _(when cultists gets their hand on wizard's magic)_ when everything can happen at the same time (cult, wiz and traitor could happen on high threat level).  

## Changelog
:cl:
add: Ported dynamic mode from /vg/, originally made by @DeityLink, @Kurfursten and @ShiftyRail
/:cl:

TheChosenEvilOne committed:
Ports dynamic gamemode from /vg/

plapatin commented:
thread theme:
https://www.youtube.com/watch?v=GXFSK0ogeg4&

Detective-Google commented:
Man, I hope we at least get a testmerge once this is ready, I've been wondering how this would work on TG

Hell maybe if we all hold hands and dream it will stop antag rolling

Lol no it won't but it'll be interesting

Fikou commented:
thread theme 2:
https://www.youtube.com/watch?v=SkTt9k4Y-a8

JJRcop commented:
@AnturK 
https://github.com/tgstation/tgstation/projects/4

Ziiro commented:
Huge fucking 👍 for this. Godspeed.

OverlyGeneric commented:
I saw someone suggest this on the Goonstation Forums. 

I hope the best of luck to you. This is a nigh Herculean task for one man.

wesoda25 commented:
Ok but what does it do

Fikou commented:
> Ok but what does it do

if its very similar to vg, roundstarts rolls a dice that makes a threat level and has a threat statistic equal to the amount of threat level
then the threat is used to roll antags, each antag costs different amounts of threat
when enough threat is generated the next latejoin person or someone from ghosts maybe chosen for an antag like ninja or wiz

Qustinnus commented:
shouldnt this use datum/gamemode instead

knacker48 commented:
Will this affect all game modes or will it be a game mode itself? Cause I for one kinda like the slower more chill rounds every now and then and would hate if they were just gone because of this

GunHog commented:
This could be trailed as its own mode, which can mix things up and provide for interactions among the antags present. Its weight against the other modes could then be adjusted according to popularity.

TheChosenEvilOne commented:
@knacker48 this is a game mode itself and this can have chill rounds if the threat level generated is under 10

TheChosenEvilOne commented:
@Qustinnus if you haven't noticed dynamic is /datum/game_mode in dynamic.dm

Qustinnus commented:
I ment dynamic game mode should be in the parent

YakumoChen commented on code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:
```diff
@@ -0,0 +1,170 @@
+	name = "Provocateur"
+	antag_datum = /datum/antagonist/rev/head
+	antag_flag = ROLE_REV_HEAD
+	restricted_roles = list("AI", "Cyborg", "Security Officer", "Warden", "Detective", "Head of Security", "Captain", "Head of Personnel", "Chief Engineer", "Chief Medical Officer", "Research Director", "Internal Affairs Agent")
```
Is this supposed to be "Lawyer" and not IAA?

TheChosenEvilOne replied:
Leftover from /vg/, removing

TheChosenEvilOne committed:
stuff or something

TheChosenEvilOne committed:
no IAA here, go away

TheChosenEvilOne committed:
largely removed all message_admins calls and removed other distribution modes

TheChosenEvilOne committed:
useless variable just for admin message, nice

TheChosenEvilOne committed:
removed simulation procs, classic_secret var and did game mode panel things

ShiftyRail commented:
Hi. I wrote a fair bit of the dynamic mode code you're porting
Not really anal about it, but if you could give credits to the main coders of it (@DeityLink, @Kurfursten and myself), that would be awesome :+1: 

Dynamic is both a blessing & a curse inasmuch as while it allows for more varied rounds, it can be extremely difficult to balance properly until the playerbase get the "right" feeling.
I can answer some code-related questions if you want, the code is sound but some parts may require explanation.

Kurfursten commented:
I actually think dynamic might work better for TG because the ability to *always* have antagonists rolling seems to fit their culture a little better (with respect).

Naturally you'll need to port your own antags to the rulesets but it's not too difficult once you understand the flow (I did a non-exhaustive readme I recommend you check out)

ShiftyRail commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,522 @@
+	for(var/datum/dynamic_ruleset/roundstart/rule in executed_rules)
+		if(rule.execute())
+			continue
+		else
```
Else what

TheChosenEvilOne replied:
huh, I were removing some of the admin message spam and I apparently didn't notice that the if-else was there because of admin messaging.

TheChosenEvilOne commented:
> Hi. I wrote a fair bit of the dynamic mode code you're porting
> Not really anal about it, but if you could give credits to the main coders of it (@DeityLink, @Kurfursten and myself), that would be awesome +1
> 
> Dynamic is both a blessing & a curse inasmuch as while it allows for more varied rounds, it can be extremely difficult to balance properly until the playerbase get the "right" feeling.
> I can answer some code-related questions if you want, the code is sound but some parts may require explanation.
> 
> Additionally, there has been some improvement to the way threat is generated since you ported it ; may want to look into that.

I added you, DeityLink and Kurfursten to the changelog.
Also can you point me towards those threat improvements? I can't seem to find them. 

> I actually think dynamic might work better for TG because the ability to _always_ have antagonists rolling seems to fit their culture a little better (with respect).
> 
> Naturally you'll need to port your own antags to the rulesets but it's not too difficult once you understand the flow (I did a non-exhaustive readme I recommend you check out and maybe port)

Yeah and sure I could port the readme.

TheChosenEvilOne commented:
uhh, that was a misclick but it is starting to be ready for review.

TheChosenEvilOne committed:
readd rigged_roundstart, add readme and remove latejoin threat generation

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_readme.dm:
```diff
@@ -0,0 +1,66 @@
+/*
```
The readme hasn't been updated to include the changes done in this PR, that must be done later.
Only change done to it was remove the faction mentions.

TheChosenEvilOne committed:
Injection cooldown changes from /vg/

ShiftyRail commented:
I was thinking mainly of https://github.com/vgstation-coders/vgstation13/pull/22646, which allows admin to select a few options before roundstart. 

![image](https://user-images.githubusercontent.com/31417754/59983272-f0d1a580-961d-11e9-9604-20377ebd95c5.png)

DeityLink commented:
:eyes:

optimumtact commented:
@Jordie0608 you want to cast your eye over this from a datum gamemode perspective

TheChosenEvilOne committed:
Merge remote-tracking branch 'origin/master' into dynamic-because-yes

TheChosenEvilOne committed:
Merge remote-tracking branch 'origin/master' into dynamic-because-yes

TheChosenEvilOne committed:
fuck I am tired

TheChosenEvilOne committed:
what the fuck

TheChosenEvilOne committed:
remove space ninja latejoin ruleset and remove duplicate code

optimumtact commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,563 @@
+			latejoin_rules = remove_rule(latejoin_rules,latejoin_rule.type)
+		spend_threat(latejoin_rule.cost)
+		threat_log += "[worldtime2text()]: Latejoin [latejoin_rule.name] spent [latejoin_rule.cost]"
+		if (latejoin_rule.execute())//this should never fail since ready() returned 1
```
add an else + crash or stack_trace call indicating what rules failed, because a comment, suprisingly, often causes failures

TheChosenEvilOne replied:
I did else + stack_trace _and comments do stop failures if you comment out the code that fails._

optimumtact commented:
`Dynamic is both a blessing & a curse inasmuch as while it allows for more varied rounds, it can be extremely difficult to balance properly until the playerbase get the "right" feeling.`
@ShiftyRail can you expand on what you mean by this?

@TheChosenEvilOne Im not opposed to trying this out.

I currently have one question, how often does this inject? Our high pop servers can have an average round length of 40-50 minutes and lower sometimes, does it inject at speeds relative to that or is it designed for lowerpop / longer rounds? (Which we still have room for on a lot of our servers)

TheChosenEvilOne committed:
fix a comment and add stack_trace calls if rules fail to execute

TheChosenEvilOne commented:
@optimumtact 
The injection for midround is every 15-50 minutes and for latejoin every 5-30 minutes (see the defines in dynamic.dm). These do need to be tweaked but not sure how much.

TheChosenEvilOne committed:
Make comments more consistent

Detective-Google commented:
This will hopefully stop people calling the shuttle just to roll antag again, since a long round *also* has a chance to spawn new antags, of course assuming it makes it that long...

DeityLink commented:
@optimumtact the injection cooldowns can be tweaked in the game mode's vars.

TheChosenEvilOne commented:
I will make the injection delays changeable by the admins.

TheChosenEvilOne committed:
Made injection delays changeable at roundstart

TheChosenEvilOne committed:
Add admin only rulesets for clockcult, overthrow, clown ops, devil and monkey.

TheChosenEvilOne committed:
admin only meteor ruleset and fixes

Krysonism commented:
PR of the year.

Our most talented coders should help out if needed, this could be great.

TheChosenEvilOne committed:
math now uses . and meteordelay in meteor is correct

TheChosenEvilOne committed:
lists and stuff

TheChosenEvilOne committed:
fixes to the panels, other things I can't remember right now

TheChosenEvilOne commented:
@Krysonism
I don't really need help, but I do need reviews so I know what to work on/fix.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets.dm:
```diff
@@ -0,0 +1,187 @@
+		if (!(antag_name in P.client.prefs.be_special) || is_banned_from(P.ckey, list(antag_name, ROLE_SYNDICATE)) || (antag_flag_override && is_banned_from(P.ckey, list(antag_flag_override, ROLE_SYNDICATE))))//are they willing and not antag-banned?
+			candidates.Remove(P)
+			continue
+		if ((exclusive_roles.len > 0) && !(P.mind.assigned_role in exclusive_roles)) // Is the rule exclusive to their job?
```
This does not actually know what role the player is going to have.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets.dm:
```diff
@@ -0,0 +1,187 @@
+		if (!(antag_name in P.client.prefs.be_special) || is_banned_from(P.ckey, list(antag_name, ROLE_SYNDICATE)) || (antag_flag_override && is_banned_from(P.ckey, list(antag_flag_override, ROLE_SYNDICATE))))//are they willing and not antag-banned?
+			candidates.Remove(P)
+			continue
+		if ((exclusive_roles.len > 0) && !(P.mind.assigned_role in exclusive_roles)) // Is the rule exclusive to their job?
```
Same thing here I think.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets.dm:
```diff
@@ -0,0 +1,187 @@
+		var/job_check = 0
+		if (enemy_roles.len > 0)
+			for (var/mob/M in mode.candidates)
+				if (M.mind && M.mind.assigned_role && (M.mind.assigned_role in enemy_roles) && (!(M in candidates) || (M.mind.assigned_role in restricted_roles)))
```
Same here, this is called before anyone knows their role.

TheChosenEvilOne replied:
this is the only blocking issue left.

TheChosenEvilOne reviewed (commented):
I am just noting things I need to change tomorrow.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -0,0 +1,761 @@
+
+	return TRUE
+	
+/datum/dynamic_ruleset/roundstart/delayed/revs/rule_process()
```
So fourth time writing this, lost all other three tries because I am writing this on my phone. The mode should have a check finished thing and check finished on highlander rulesets then call the round report from the ruleset that ended the round

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -0,0 +1,761 @@
+	high_population_requirement = 101
+
+/datum/dynamic_ruleset/roundstart/extended/pre_execute()
+	message_admins("Starting a round of extended.")
```
Maybe make this spend all of the threat.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -0,0 +1,761 @@
+	message_admins("Starting a round of extended.")
+	log_admin("Starting a round of extended.")
+	return TRUE
+
```
All rulesets beyond this point are untested except for meteor.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -0,0 +1,761 @@
+//                                          //
+//////////////////////////////////////////////
+
+/datum/dynamic_ruleset/roundstart/delayed/revs
```
Not sure how the delay affects the giving of the flash.

TheChosenEvilOne replied:
it is actually fine, if you don't have space you just don't get one.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -0,0 +1,761 @@
+	main_cult.setup_objectives()
+	return TRUE
+
+/datum/dynamic_ruleset/roundstart/bloodcult/round_result()
```
See the comment for revs rule_process.

TheChosenEvilOne reviewed (commented):
I have to do these file by file because I am on mobile. This is pain.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:
```diff
@@ -0,0 +1,132 @@
+/datum/dynamic_ruleset/latejoin/provocateur/execute()
+	var/mob/M = pick(candidates)
+	assigned += M
+	spawn(1 SECONDS)
```
This is unneeded I think.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:
```diff
@@ -0,0 +1,132 @@
+//                                          //
+//////////////////////////////////////////////
+
+/datum/dynamic_ruleset/latejoin/wizard
```
I am not actually sure if this still announces they have signed up.

TheChosenEvilOne replied:
ill just remove the wizard latejoin ruleset for now.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:
```diff
@@ -0,0 +1,132 @@
+			return FALSE
+	return ..()
+
+/datum/dynamic_ruleset/latejoin/execute()
```
Should add special role I guess.

TheChosenEvilOne reviewed (commented):
Nothing much to say about this.

TheChosenEvilOne committed:
added check_finished to dynamic, removed latejoin wizard ruleset, fixed latejoin rev ruleset and others.

TheChosenEvilOne committed:
Nearly done

TheChosenEvilOne commented:
This would be ready if it wasn't for the enemy role checking, this is only an issue on round start rulesets but still. The issue is that ready checks if there are enough enemies for a rule to be executed but at that time assigned_role isn't set yet, so every roundstart rule that requires enemies (everything except for traitor one and some at high threat) fails at this point.

TheChosenEvilOne committed:
compile fix

TheChosenEvilOne committed:
now it should work.

TheChosenEvilOne committed:
custom intercept report

TheChosenEvilOne committed:
fixed the linter issue?

TheChosenEvilOne committed:
Removed enemy_jobs from roundstart rulesets.

TheChosenEvilOne commented:
> 
> 
> This would be ready if it wasn't for the enemy role checking, this is only an issue on round start rulesets but still. The issue is that ready checks if there are enough enemies for a rule to be executed but at that time assigned_role isn't set yet, so every roundstart rule that requires enemies (everything except for traitor one and some at high threat) fails at this point.

So I just decided to axe enemy_roles from roundstart rulesets. It's not like the amount of security officers was checked before.

TheChosenEvilOne committed:
fixes some issues that were found.

TheChosenEvilOne committed:
This seems to have fixed sec getting antag.

TheChosenEvilOne committed:
Ruleset requirement changes, removed useless proc.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Made malfunctioning AI not a highlander ruleset.

TheChosenEvilOne committed:
Injection cooldowns are now counted as minutes.
Slightly decreased the threat required for more roundstart rulesets.
Forcing injection now forces injection.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Changes "log_admin"s to "log_game"s.
This was done mainly because of statbus.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
prefix log messages with "DYNAMIC:".
Added more logging.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Made set_round_result call check_finished on rules
Makes sure that only the rule that ended the round sets result.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed dead players being given antag.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

FloranOtten commented:
Will this enable the Dread combo of Nukies, Cult and Wizard?

TheChosenEvilOne commented:
> 
> 
> Will this enable the Dread combo of Nukies, Cult and Wizard?

mayhaps.

FloranOtten commented:
> > Will this enable the Dread combo of Nukies, Cult and Wizard?
> 
> mayhaps.

FUCKING SOLD

TheChosenEvilOne committed:
Fixed some issues found.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

Rowelll commented:
merge

Arathianen commented:
This wasted tested today on Terry. It was a fucking chaotic mess that no one could escape alive from. The station crashed and burned. Everybody died. It was pure fucking chaos.

Most fun shit ever. Speedmerge it.

TheChosenEvilOne committed:
Fixed issues with midround rulesets.
Added blood brother ruleset (UNTESTED).

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
I am stupid, this created issues.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

dee-dubya commented:
merge it chief

AnturK reviewed (changes requested):
I'll finish reading this over when i get home.

TheChosenEvilOne committed:
special_role is now assigned at the correct time.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Add a chance (30%) for changeling team mode.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed compiling error.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets.dm:
```diff
@@ -0,0 +1,194 @@
+	if(repeatable && weight > 1)
+		for(var/datum/dynamic_ruleset/DR in mode.executed_rules)
+			if(istype(DR,src.type))
+				weight = max(weight-2,1)
```
Weight is decreased if the ruleset is repeatable so the more it is repeated the less it appears.

TheChosenEvilOne replied:
probably will make this configurable though

TheChosenEvilOne replied:
Made this configurable but I don't think the variable name is that good.

TheChosenEvilOne committed:
Requested changes and fixes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,628 @@
+	var/list/roundstart_rules = list()
+	var/list/latejoin_rules = list()
+	var/list/midround_rules = list()
+	var/list/second_rule_req = list(100,100,80,70,60,40,20,0,0,0)// Requirements for extra round start rules
```
Will do later, right now focused on making it work correctly.

TheChosenEvilOne committed:
Forgot some "var/"s

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,628 @@
+			continue
+		if (M.stat != DEAD)
+			current_players[CURRENT_LIVING_PLAYERS].Add(M)
+			if (M.mind && (M.mind.special_role))
```
added a check for antag_datums len.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,628 @@
+			current_rules -= rule
+
+	if (midround_injection_cooldown > 0)
+		midround_injection_cooldown -= (world.time - last_injection_update)
```
I am not sure if I understand what you mean with this.

TheChosenEvilOne committed:
forgot this one.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

AnturK reviewed (changes requested):
Few more

TheChosenEvilOne committed:
more review changes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed latejoin assignment ignoring prefs.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Provocateur latejoin antag_flag is now correct.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,628 @@
+	var/list/current_rules = list()
+	var/list/executed_rules = list()
+
+	var/list/list/current_players = list(CURRENT_LIVING_PLAYERS, CURRENT_LIVING_ANTAGS, CURRENT_DEAD_PLAYERS, CURRENT_OBSERVERS)
```
helps to keep track of players which might benefit future rulesets, but I can change it. As well as when debugging this was really helpful.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets.dm:
```diff
@@ -0,0 +1,194 @@
+		if(P.mind.special_role) // We really don't want to give antag to an antag.
+			candidates.Remove(P)
+			continue
+		if (!(antag_name in P.client.prefs.be_special) || is_banned_from(P.ckey, list(antag_name, ROLE_SYNDICATE)) || (antag_flag_override && is_banned_from(P.ckey, list(antag_flag_override, ROLE_SYNDICATE))))//are they willing and not antag-banned?
```
not sure what you mean with post SQL validation but I added age checks.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,628 @@
+		if (initial(DR.weight))
+			midround_rules += new rule()
+	for(var/mob/dead/new_player/player in GLOB.player_list)
+		if(player.ready == PLAYER_READY_TO_PLAY && player.mind)
```
You mean validation if they even have anything enabled in preferences for candidates? This is just used for counting roundstart pop and candidates which are later filtered by preferences by the ruleset.

TheChosenEvilOne committed:
Added ages checks and removed /

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
I really should start compiling before pushing.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Made repeatable weight decrease configurable.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Removed the unused starting_threat variable.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

optimumtact commented on code/game/gamemodes/dynamic/dynamic.dm:
```diff
@@ -0,0 +1,628 @@
+		dat += "<br/>"
+		for (var/datum/dynamic_ruleset/DR in executed_rules)
+			var/ruletype = ""
+			if (istype (DR, /datum/dynamic_ruleset/roundstart))
```
`var/ruletype = DR.ruletype`

TheChosenEvilOne committed:
Midround nuke ops now assigns nuke_team.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Midround ghost rules now set special_role.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
High pop check now checks playing population.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
last_injection_update is not used anymore.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
I can't type.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

jegub commented:
Someone was made a headrev on round 113281 on Terry, they have revolutionary disabled in their game preferences and their flash was not converting anyone.

TheChosenEvilOne commented:
> 
> 
> Someone was made a headrev on round 113281 on Terry, they have revolutionary disabled in their game preferences and their flash was not converting anyone.

No idea about the flash, but people getting antagonist from latejoin when it is disabled has been fixed. This was an issue I fixed yesterday.

TheChosenEvilOne committed:
Reviews, acceptable change.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Move midround wizard to wizard spawn point.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheRespectedMan commented:
@TheChosenEvilOne 
![image](https://user-images.githubusercontent.com/22877546/61325540-aa86f500-a81d-11e9-9ce8-6f214e5af3fa.png)

sometimes it leads to redundancy

TheChosenEvilOne committed:
Fixed traitor process and commenting.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed inverted age_check.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne commented:
> 
> 
> @TheChosenEvilOne
> ![image](https://user-images.githubusercontent.com/22877546/61325540-aa86f500-a81d-11e9-9ce8-6f214e5af3fa.png)
> 
> sometimes it leads to redundancy

I were observing that round and I knew of this from an earlier round, not sure if this has been fixed but I think it has been.

TheChosenEvilOne committed:
Runtimes are bad.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Something something, I am stupid.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed blood brother.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed successful midround triggering a new one

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Ruleset fixes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
woah, I still don't compile before pushing.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Actually fixed double antag issue.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Rules with no check_finished now set round result.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

FloranOtten commented:
Have noticed that certain gamemodes are not at all balanced for this. Having cult and changelings and traitors about, when each of those was balanced as a gamemode unto itself, makes the game extremely chaotic.

TheChosenEvilOne committed:
Fixes and other changes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
oh wait, this does not work.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
aaaaaaaaaaaaaaaaaaa

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Width changing works now.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Round end threat and executed rules + fixes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

ninjanomnom commented on code/game/gamemodes/dynamic/dynamic_maths.dm:
```diff
@@ -0,0 +1,48 @@
+// Used for midround/latejoin injection delays.
+/proc/exp_distribution(var/desired_mean)
+	if (desired_mean <= 0)
+		desired_mean = 1
```
I feel you shouldn't be hiding potential errors from people passing in 0 to this

ninjanomnom commented on code/game/gamemodes/dynamic/dynamic_maths.dm:
```diff
@@ -0,0 +1,48 @@
+// Exponential distribution.
+// Used for midround/latejoin injection delays.
+/proc/exp_distribution(var/desired_mean)
```
`var` is unnecessary in proc arguments
```suggestion
/proc/exp_distribution(desired_mean)
```

ninjanomnom replied:
This whole proc could be replaced with a define and put into the math helpers
`#define EXP_DISTRIBUTION(desired_mean) ( -(1/(1/desired_mean)) * log(rand(1, 1000) * 0.001) )`

ninjanomnom commented on code/game/gamemodes/dynamic/dynamic_maths.dm:
```diff
@@ -0,0 +1,48 @@
+	var/lambda = 1/desired_mean
+	var/x = rand()
+	while (x == 1)
+		x = rand()
```
Drop the while and do something like
`rand(0, 999) * 0.00.1` or however many significant digits you want

ninjanomnom commented on code/game/gamemodes/dynamic/dynamic_maths.dm:
```diff
@@ -0,0 +1,48 @@
+
+// Lorentz/Cauchy distribution.
+// Used for generating threat.
+/proc/lorentz_distribution(var/x0, var/s)
```
another define

ninjanomnom commented on code/game/gamemodes/dynamic/dynamic_maths.dm:
```diff
@@ -0,0 +1,48 @@
+
+// Cumulative Lorentz/Cauchy distribution.
+// Used for generating peaceful percentage.
+/proc/lorentz_cummulative_distribution(var/x, var/x0, var/s)
```
another define

TheChosenEvilOne committed:
Reviews, moved game mode panel to game panel.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed roundstart rules not being accepted.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

Cyprex commented:
View Log for threat level in the game mode panel does nothing

TheChosenEvilOne committed:
Fixed threat log not working.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed forced injection.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Return some admin messaging.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

DamianX commented:
dynamic was a mistake

DeityLink commented:
nay, the only mistake was neutering it.

Cyprex commented on code/game/gamemodes/dynamic/dynamic_rulesets.dm:
```diff
@@ -0,0 +1,209 @@
+
+	var/flags = 0
+
+	// For midround polling
```
// Stores candidates that hit yes on a midround prompt
Is candidates the right word here?

TheChosenEvilOne replied:
I will move this to midround rulesets and I got a comment locally for this already.

Cyprex replied:
I see, that certainly makes sense!

ShiftyRail commented:
>dynamic ~~was~~ *is* a mistake

MadmanMartian commented:
have fun

TheChosenEvilOne committed:
Commenting, fixes + p e r f o r m a n c e

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Allow delayed rulesets to be ran.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

Cyprex commented:
Round ID 113558 on Terry, 53 threat yet no rulesets executed. The following round, with similiar threat, no roundstart rulesets were executed. Forcing a midround nuclear assault worked, although only 2 operatives were chosen due to a lack of applicants. There were a total of 45 players connected at the time.

Cyprex commented:
Addendum: The nuclear assault did not get set up properly, no leader was chosen, preventing the ops from getting onto their shuttle.

Cyprex commented:
Another round: "Syndicate Sleeper Agent",triggered 4 times manuall, only created 1 traitor each. Nothing happened when I hit "Now!" on the injection timers, only "Execute midround ruleset" worked!

TheChosenEvilOne committed:
Revert "p e r f o r m a n c e"

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne commented:
I have now fixed this issue, this is why you test your changes people!

Cyprex commented:
A midround traitor with the "Steal a silver of the Supermatter Crystal" Objective spawned without the box containing tools for it. Round ID 113600 on Terry.

MadmanMartian commented:
We're having some issues with roundstart rules not firing despite being elligible. Will keep you posted when we have a fix.

DeityLink commented:
aaaaaand just like that we got the fix. it was due to a PR from two weeks ago so its code may not even be in TG's branch. #23615

TheChosenEvilOne commented:
@Cyprex
This is not an issue on the side of dynamic but ill check if i can make it work. The box most likely is given only on roundstart

Cyprex commented:
@DeityLink I believe you meant to link https://github.com/vgstation-coders/vgstation13/pull/23615

DeityLink commented:
that I did

TheChosenEvilOne committed:
Fixed a division by zero error.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed midround ghost index being inverted.
Fixed rules with no candidates being executed.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
merge conflict

TheChosenEvilOne committed:
Actually fixed the division by zero error.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Commenting, renamed readme file

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Commenting and other small fixes.
Removed overthrow ruleset.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
save - work

TheChosenEvilOne committed:
Blob midround

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Merge branch 'dynamic-because-yes' of github.com:TheChosenEvilOne/tgstation into dynamic-because-yes

TheChosenEvilOne committed:
Removed yesp.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Add threat level forcing and fixed revs.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Moved midround cooldown assigning.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
More fixes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Logging.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

kriskog commented:
#45405

TheChosenEvilOne commented:
> 
> 
> #45405 bug for ya

Fixed in commit 13266be

TheChosenEvilOne committed:
Removed hivemind.
Commented out revs for technical issues.
Added threat cost to roundend.
Added blocking ruleset support.
Fixed midround rulesets "failing to execute" even when successful.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

bigfatbananacyclops commented:
dynamic is great but there's a few antags that dont really collide with others, operatives tend to work with agents and it will multiply their forces, revs converting antags will give them a strong fighter, same with cult. i honestly think if it picks ops, cult or revs, it should just pick one of them, but it should also however, be extremely rare to pick another antag with them. so the rounds cant really be metad much, expect anything.

Cyprex commented:
I feel like it would be better(outside of the scope of this PR though) by granting certain antags conversion immunity, it's silly that a wizard can just get flashed by a revhead and converted in the process for example.

TheChosenEvilOne committed:
Code improvement.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Decreases the chance of a wizard.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

Akrilla commented:
Okay so I hear this literally 24/7 on Terry and Discord. Wizards are way too damn common, something that's meant to be once in a blue moon I'm seeing about 3 times a day when I play.

TheChosenEvilOne commented:
> 
> 
> Okay so I hear this literally 24/7 on Terry and Discord. Wizards are way too damn common, something that's meant to be once in a blue moon I'm seeing about 3 times a day when I play.

Chance recently decreased

Akrilla commented:
based

TheChosenEvilOne committed:
Enabled revs with a different approach.
Fixed game mode panel showing wrong time for injections.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Tweaked rule requirements and pop per requirement.
Added nightmare and xenomorph rulesets.
Tweaked blood brothers ruleset.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

BeeSting12 commented:
Server-Side Code Modifications: https://tgstation13.org/parsed-logs/basil/server_side_modifications.dm


Round ID: 114557
BYOND Version: 512.1471
Server revision compiled on: 
Master commit: 1d252de949178a8caf2b5d38f23f4bbb8e79f10a
The following pull requests are currently test merged:
#44639: '[Ready] Ports dynamic gamemode from /vg/' by TheChosenEvilOne at commit fccb0c5b97
#43148: '[s]Temp workaround for admins matching stickybans because of a byond bug.' by MrStonedOne at commit e34fa30ce1
#45457: 'Tend Wounds Now Is -2 Steps and Borg-Friendly!' by ExcessiveUseOfCobblestone at commit b253806759
#45555: '[s]Throttle mind notes (memories)' by MrStonedOne at commit 545d99d38e

Local commit: ff79ce15429b39935711dfae6f00c1722c29c1c4
Server tools version: /tg/station 13 Server v3.2.6.2

Had a case of a posibrain getting traitor. Dont think that's intended.

TheChosenEvilOne committed:
Tweaked values.
Posibrains can no longer be traitors.
Fixed nukeops and clownops "failing" to execute.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne commented:
@BeeSting12 Fixed in the latest commit.

TheChosenEvilOne committed:
Increased the threat requirement for cult.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Tweaked cult numbers.
Forced roundstart rulesets now get candidates again.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Code improvement:
Added ONLY_RULESET flag, accepting name suggestions.
HIGHLANDER_RULESET and ONLY_RULESET checks are faster.
Game mode panel finally shows correct value.
Removed the check_for_highlander proc.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
A weird formula that scales threat with pop.
If this is bad, revert me.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Merged midround and latejoin picking procs.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
minimum_players and maximum_players for rulesets.
Commenting changes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

kriskog commented:
Round 114816
![image](https://user-images.githubusercontent.com/33292112/62332143-aa922080-b4be-11e9-98ec-76e93d2b6654.png)

Charlie station ghost role got sleeper agent traitor roll.

TheChosenEvilOne committed:
Merge remote-tracking branch 'origin/master' into dynamic-because-yes

TheChosenEvilOne committed:
Fixed ghost roles getting midrounds.
Removed useless variable.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Begone src and revs "but I give up" edition.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Fixed latejoin runtime.
Candidates list on midround now gets cleared.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
So, basically dog got traitor.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
uhhh.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
nukeops but now with lessops.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Startup is now faster.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

davethwave commented:
This stuff reminds me of when you could become a lavaland golem ling/traitor since you used to become them before the crew spawned in.

TheChosenEvilOne committed:
Logging changes.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

TheChosenEvilOne committed:
Nukeops no longer end the round when killed.

Signed-off-by: TheChosenEvilOne <tceo-email@protonmail.com>

MadmanMartian commented:
best of luck
